hypothesis-python/src/hypothesis/strategies/_internal/datetime.py

```diff
@@ -62,30 +88,50 @@ def draw_capped_multipart(data, min_value, max_value):
         result[name] = val
         cap_low = cap_low and val == low
         cap_high = cap_high and val == high
+    if hasattr(min_value, "fold"):
+        # The `fold` attribute is ignored in comparison of naive datetimes.
+        # In tz-aware datetimes it would require *very* invasive changes to
+        # the logic above, and be very sensitive to the specific timezone
+        # (at the cost of efficient shrinking and mutation), so at least for
+        # now we stick with the status quo and generate it independently.
+        result["fold"] = utils.integer_range(data, 0, 1)
     return result
 
 
 class DatetimeStrategy(SearchStrategy):
-    def __init__(self, min_value, max_value, timezones_strat):
+    def __init__(self, min_value, max_value, timezones_strat, allow_imaginary):
         assert isinstance(min_value, dt.datetime)
         assert isinstance(max_value, dt.datetime)
         assert min_value.tzinfo is None
         assert max_value.tzinfo is None
         assert min_value <= max_value
         assert isinstance(timezones_strat, SearchStrategy)
+        assert isinstance(allow_imaginary, bool)
         self.min_value = min_value
         self.max_value = max_value
         self.tz_strat = timezones_strat
+        self.allow_imaginary = allow_imaginary
 
     def do_draw(self, data):
-        result = draw_capped_multipart(data, self.min_value, self.max_value)
-        result = dt.datetime(**result)
+        # We start by drawing a timezone, and an initial datetime.
         tz = data.draw(self.tz_strat)
+        result = self.draw_naive_datetime_and_combine(data, tz)
+
+        # TODO: with some probability, systematically search for one of
+        #   - an imaginary time (if allowed),
+        #   - a time within 24hrs of a leap second (if there any are within bounds),
+        #   - other subtle, little-known, or nasty issues as described in
+        #     https://github.com/HypothesisWorks/hypothesis/issues/69
+
+        # If we happened to end up with a disallowed imaginary time, reject it.
+        if (not self.allow_imaginary) and datetime_does_not_exist(result):
```

Comment:
One thing to note: I was originally envisioning that `allow_imaginary=False` would only ever set the `fold` value if it's actually required by always returning the *result* of the round-tripping to UTC, since you're already doing that.

I don't know if that would require renaming the argument to make it clearer that that's happening, but you could achieve this by inverting the sense of `datetime_does_not_exist` and having it either return `False` (or `None`, depending on your aesthetic preferences) or the round-tripped version of the datetime, so you can do something along these lines:

```python
if not self.allow_imaginary:
    roundtripped = datetime_exists(result)
    if not roundtripped:
        data.mark_invalid()
    else:
        result = roundtripped
```

Of course, end users can always choose to do this with `.map(lambda dt: dt.astimezone(UTC).astimezone(dt.tzinfo)`, but since you've already done the work *anyway*, it might be preferable to just return this, with the implication that `allow_imaginary=False` means, "Give me a `datetime` that is the result of calling `.astimezone(tz)`."

All that said, if we assume that there are some users who want `fold` set randomly and some users who want it set only when necessary (for `allow_imaginary=False`), there is an asymmetry between what's necessary for end users to do in the "unsupported" case. With the current behavior, `allow_imaginary=False` always generates a *round-trippable* `datetime`, so you can be guaranteed that round-tripping it won't cause problems. In order to implement the current behavior in a scenario where `allow_imaginary=False` never returns a random `fold`, you'd either need to generate your own strategy that randomly sets `fold` or you'd need to implement `datetime_does_not_exist` (with its somewhat tricky `OverflowError` behavior) yourself and use it as a `filter`.

So, in the end, I'd say do whatever you think is the better interface. I doubt many people will notice in the end.

Reply:
Ah, I see what you mean... while I don't expect anyone to notice, I'm going to stick with the current random-`fold` generation, due to the asymmetry you note and because as a general principle we try to generate the least constrained values possible.  We can always restrict this later if it's an issue!